### PR TITLE
[SPARK-50117][BUILD][SS] Change to using `maven/sbt` plugin to generate `StateMessage.java`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr changes the process to dynamically generate `StateMessage.java` during build using the corresponding `maven/sbt` plugin.

### Why are the changes needed?
Maintain the same generation way as the Java files corresponding to other pb files in Spark.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Pass GitHub Actions
- All maven test passed: https://github.com/LuciferYang/spark/runs/32055792763

image

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -1166,3 +1166,9 @@ jobs:
           cd ui-test
           npm install --save-dev
           node --experimental-vm-modules node_modules/.bin/jest
+
+  maven-test:
```
Comment (LuciferYang) [inline .github/workflows/build_and_test.yml:1170]: will revert after test

**Comment by LuciferYang** [comment]
cc @dongjoon-hyun FYI

**Comment by dongjoon-hyun** [comment]
Could you resolve the conflicts, @LuciferYang ?

**Review by dongjoon-hyun — COMMENTED** [review]
cc @bogao007 , @jingz-db , @HeartSaVioR from SPARK-49821
- https://github.com/apache/spark/pull/48290

**Comment by LuciferYang** [comment]
> Could you resolve the conflicts, @LuciferYang ?

done

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM from my side. Thank you, @LuciferYang .

**Comment by HeartSaVioR** [comment]
Thanks for working on this, @LuciferYang !
I don't have a full context of how PySpark folks do the generated code thing in Spark Connect, but we also have py file being generated from proto. Will we need to do the same for generated py file? Or would there be technical reason not to do this? cc. @HyukjinKwon

**Comment by LuciferYang** [comment]
Thank you for the question from @HeartSaVioR . This was also an issue I intended to discuss in the original pr, and I apologize for forgetting to leave a comment yesterday. 

Regarding the `spark-connect` module, there is a script called `dev/connect-gen-protos.sh`. This script is used by developers to manually regenerate the corresponding `_pb2.py` and `_pb2.pyi` files and submit them when changes are made to the `.proto` files in the `connect` module. Personally, I think we should provide a similar way to automate this process in the future.

**Comment by LuciferYang** [comment]
@HeartSaVioR @dongjoon-hyun Can we merge this pr? I have created a new Jira ticket: SPARK-50139, to remind us that we should provide a tool to assist in regenerating `StateMessage_pb2.py` and `StateMessage_pb2.pyi`. We can complete this TODO after determining the most reasonable approach for PySpark.

**Comment by dongjoon-hyun** [comment]
To @LuciferYang , I agree with you.

I believe we can handle that seperately in SPARK-50139 if needed. Or, we can do a follow-up PR if something is missing in this PR's context.

**Comment by dongjoon-hyun** [comment]
If there is no further comments until Tomorrow (72 hours after https://github.com/apache/spark/pull/48654#issuecomment-2439177691), I believe you can merge this.

**Comment by LuciferYang** [comment]
> If there is no further comments until Tomorrow (72 hours after [#48654 (comment)](https://github.com/apache/spark/pull/48654#issuecomment-2439177691)), I believe you can merge this.

Got it ~

**Comment by HeartSaVioR** [comment]
Would we want to also add StateMessage.java to gitignore so that we don't try to add the file?

**Comment by LuciferYang** [comment]
> Would we want to also add StateMessage.java to gitignore so that we don't try to add the file again?

double checked:


- Maven

```
build/mvn clean install -DskipTests -pl sql/core -am


[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spark Project Parent POM 4.0.0-SNAPSHOT:
[INFO] 
[INFO] Spark Project Parent POM ........................... SUCCESS [  3.451 s]
[INFO] Spark Project Tags ................................. SUCCESS [  5.497 s]
[INFO] Spark Project Sketch ............................... SUCCESS [  5.819 s]
[INFO] Spark Project Common Utils ......................... SUCCESS [ 26.619 s]
[INFO] Spark Project Local DB ............................. SUCCESS [ 13.609 s]
[INFO] Spark Project Networking ........................... SUCCESS [ 16.461 s]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [ 15.016 s]
[INFO] Spark Project Variant .............................. SUCCESS [  4.177 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [ 14.669 s]
[INFO] Spark Project Connect Shims ........................ SUCCESS [  7.762 s]
[INFO] Spark Project Launcher ............................. SUCCESS [  7.345 s]
[INFO] Spark Project Core ................................. SUCCESS [02:26 min]
[INFO] Spark Project SQL API .............................. SUCCESS [ 40.995 s]
[INFO] Spark Project Catalyst ............................. SUCCESS [03:04 min]
[INFO] Spark Project SQL .................................. SUCCESS [03:19 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11:33 min
[INFO] Finished at: 2024-10-29T08:28:02+08:00
[INFO] ------------------------------------------------------------------------


git status

On branch state_message-pb
Your branch is up to date with 'origin/state_message-pb'.

nothing to commit, working tree clean


find sql/core -name "*StateMessage.java"
sql/core/target/generated-sources/org/apache/spark/sql/execution/streaming/state/StateMessage.java
```


- SBT

```
build/sbt clean "sql/compile"
[success] Total time: 109 s (01:49), completed 2024年10月29日 上午8:34:33

git status
On branch state_message-pb
Your branch is up to date with 'origin/state_message-pb'.

nothing to commit, working tree clean

find sql/core -name "*StateMessage.java"
sql/core/target/scala-2.13/src_managed/main/org/apache/spark/sql/execution/streaming/state/StateMessage.java
```

The `StateMessage.java` files generated by the maven/sbt plugin are all located in the `sql/core/target` directory. They are all visible on the classpath and have already been made invisible to git. I believe that under normal circumstances, they will not be accidentally added by `git add`, so there is no need to modify the `.gitignore` file any further.

**Comment by HeartSaVioR** [comment]
Ah OK, so the generated file is placed in target, not source directory. Sounds good.

**Comment by HeartSaVioR** [comment]
Please allow me to wait for a day to sync with folks working on this (TWS in PySpark). I've asked them to comment and they will come in US tz. If there is no comment in tomorrow US tz, I'm OK with the fix.

**Comment by LuciferYang** [comment]
> Please allow me to wait for a day to sync with folks working on this (TWS in PySpark). I've asked them to comment and they will come in US tz. If there is no comment in tomorrow US tz, I'm OK with the fix.

OK ~

**Comment by dongjoon-hyun** [comment]
Thank you for the sync-up with them, @HeartSaVioR .

**Comment by dongjoon-hyun** [comment]
To @bogao007 , yes, exactly.

FYI, Apache Spark has many `proto` files (including this) and use them (except this) in that way.
```
$ find . -type f -name "*.proto" | grep -v test
./core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto
./sql/core/src/main/java/org/apache/spark/sql/execution/streaming/StateMessage.proto
./sql/connect/common/src/main/protobuf/spark/connect/relations.proto
./sql/connect/common/src/main/protobuf/spark/connect/catalog.proto
./sql/connect/common/src/main/protobuf/spark/connect/base.proto
./sql/connect/common/src/main/protobuf/spark/connect/example_plugins.proto
./sql/connect/common/src/main/protobuf/spark/connect/types.proto
./sql/connect/common/src/main/protobuf/spark/connect/expressions.proto
./sql/connect/common/src/main/protobuf/spark/connect/commands.proto
./sql/connect/common/src/main/protobuf/spark/connect/common.proto
```

**Comment by LuciferYang** [comment]
> Thanks for working on this change! QQ: for the dev loop after this is merged, if we add changes to `StateMessage.proto`, would command `build/sbt clean package` be enough to automatically generate the java file and get picked up by the compiler?

@bogao007  Yes, it would be enough.

Thanks for your help! @dongjoon-hyun

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by LuciferYang** [comment]
Merged into master for Spark 4.0. Thanks @dongjoon-hyun @HeartSaVioR and @bogao007

**Comment by dongjoon-hyun** [comment]
Thank you all again!

**Pull request closed without merging**